### DISCUSSION TITLE
How to pass a selected option to the constructor

### DISCUSSION THREAD

**Discussion by jurekms** [remkop/picocli#1522] [Q&A]
Hi

I really like picocli. I am currently creating a program to manage the water meter. Communication is carried out using the modbus serial or TCP/IP protocol. I wanted to provide the application with information about the selected protocol (-RS or -TCP). The appropriate connection is retrieved in the Constructor of the WaterMeterCLI class. When the constructor is run, there is no static connectionOptions object, which means that I can not determine what connection to use. 

regards Jurek

'	package pl.rynekdebnicki.waterMeter;

	import java.util.concurrent.Callable;

	import com.ghgande.j2mod.modbus.facade.AbstractModbusMaster;

	import picocli.CommandLine.ArgGroup;
	import picocli.CommandLine.Command;
	import picocli.CommandLine.HelpCommand;
	import picocli.CommandLine.Option;

	@Command(name = "waterMeter", subcommandsRepeatable = true, subcommands = HelpCommand.class)
	public class WaterMeterCLI implements Callable<Integer>  {
		WaterMeterConnection wmConnection;
		WaterMeter waterMeter;
		AbstractModbusMaster master;

		@ArgGroup(exclusive = true)	
		ConnectionOptions connectionOptions;
		static class ConnectionOptions {
			@Option(names = {"-TCP"}) boolean connectionTCP;
			@Option(names = {"-RS"})  boolean connectionRS;
		}

		static class RedLedOptions {
			@Option(names = {"-redON"}) boolean redLedON;
			@Option(names = {"-redOFF"}) boolean redLedOFF;
		}
		
		static class GreenLedOptions {
			@Option(names = {"-greenON"}) boolean greenON;
			@Option(names = {"-greenOFF"}) boolean greenOFF;
		}
		
		@Command(name = "Led")
		void Led(	@ArgGroup(exclusive = true, multiplicity = "1")	RedLedOptions redLedOptions, 
					@ArgGroup(exclusive = true, multiplicity = "1")	GreenLedOptions greenLedOptions
				) {
			
			if(redLedOptions.redLedON) {
				waterMeter.SetRedLed();
			}
			
			if(redLedOptions.redLedOFF) {
				waterMeter.ResetRedLed();
			}
			
			if(greenLedOptions.greenON) {
				waterMeter.SetGreenLed();
			}
			
			if(greenLedOptions.greenOFF) {
				waterMeter.ResetGreenLed();
			}
		}
		

		
		public WaterMeterCLI() {
			System.out.println("Podłaczenie do urzadzenia MODBUS...");
			wmConnection = new WaterMeterConnection();
			
			
			if(connectionOptions.connectionRS) {
				master = wmConnection.getRSModbusMaster();
			}
			
			if(connectionOptions.connectionTCP) {
				master = wmConnection.getTCPModbusMaster();
			}
					
			waterMeter = new WaterMeter(master, wmConnection.getWmConnectionConf());
		}
		
		
		@Override
		public Integer call() throws Exception {
			return 1;
		}

	}

`

I found some solution :)

`

	private void init() {
		System.out.println("Podłaczenie do urzadzenia MODBUS...");
		wmConnection = new WaterMeterConnection();
		AbstractModbusMaster master = null;
		if (connectionOptions.connectionRS) {
			master = wmConnection.getRSModbusMaster();
		}

		if (connectionOptions.connectionTCP) {
			master = wmConnection.getTCPModbusMaster();
		}
		waterMeter = new WaterMeter(master, wmConnection.getWmConnectionConf());

	public int executionStrategy(ParseResult parseResult) {
		init(); 
		return new CommandLine.RunLast().execute(parseResult); 
	}
`

**Comment by remkop** [comment]
Yes, that is one idea.
Another idea is to just have one command instead of trying to configure the top-level command with a subcommand.
For example:

```java
@Command(name = "waterMeter", subcommandsRepeatable = true, subcommands = HelpCommand.class)
public class WaterMeterCLI implements Callable<Integer> {
    WaterMeterConnection wmConnection;
    WaterMeter waterMeter;
    AbstractModbusMaster master;

    @ArgGroup(exclusive = true) ConnectionOptions connectionOptions;
    @ArgGroup(exclusive = true, multiplicity = "1") RedLedOptions redLedOptions;
    @ArgGroup(exclusive = true, multiplicity = "1") GreenLedOptions greenLedOptions;

    static class ConnectionOptions {
        @Option(names = {"-TCP"}) boolean connectionTCP;
        @Option(names = {"-RS"})  boolean connectionRS;
    }

    static class RedLedOptions {
        @Option(names = {"-redON"}) boolean redLedON;
        @Option(names = {"-redOFF"}) boolean redLedOFF;
    }

    static class GreenLedOptions {
        @Option(names = {"-greenON"}) boolean greenON;
        @Option(names = {"-greenOFF"}) boolean greenOFF;
    }

    @Override
    public Integer call() throws Exception {
        System.out.println("Podłaczenie do urzadzenia MODBUS...");
        wmConnection = new WaterMeterConnection();
        if(connectionOptions.connectionRS) { master = wmConnection.getRSModbusMaster();}
        if(connectionOptions.connectionTCP) { master = wmConnection.getTCPModbusMaster();}

        waterMeter = new WaterMeter(master, wmConnection.getWmConnectionConf());

        if(redLedOptions.redLedON) { waterMeter.SetRedLed();}
        if(redLedOptions.redLedOFF) { waterMeter.ResetRedLed(); }
        if(greenLedOptions.greenON) { waterMeter.SetGreenLed(); }
        if(greenLedOptions.greenOFF) { waterMeter.ResetGreenLed(); }

        // remaining business logic here...
        return 1;
    }

    public static void main(String... args) {
        int exitCode = new CommandLine(new WaterMeterCLI()).execute(args);
    }
}
```
